dmoj/judge.py

```diff
@@ -67,239 +93,377 @@ def _updater_thread(self):
+        # FIXME(tbrindus): marked Any pending grader cleanups.
+        self.grader: Any = None
+
+        self.worker_process_conn, child_conn = multiprocessing.Pipe()
+        self.worker_process = multiprocessing.Process(
+            name="DMOJ Judge Handler for %s/%d" % (self.submission.problem_id, self.submission.id),
+            target=self._worker_process_main,
+            args=(child_conn, self.worker_process_conn),
+        )
+        self.worker_process.start()
+        child_conn.close()
+
+    def communicate(self) -> Generator[Tuple[IPC, tuple], None, None]:
+        while True:
+            try:
+                ipc_type, data = self.worker_process_conn.recv()
+            except Exception:
+                logger.error("Failed to read IPC message from worker!")
+                raise
+
+            if ipc_type == IPC.BYE:
+                self.worker_process_conn.send((IPC.BYE, ()))
+                return
+            else:
+                yield ipc_type, data
+
+    def stop(self) -> None:
+        if self.worker_process and self.worker_process.is_alive():
+            # Might be None if run was never called, or failed.
+            try:
+                self.worker_process.join(timeout=IPC_TEARDOWN_TIMEOUT)
+            except OSError:
+                logger.exception('Exception while waiting for worker to shut down, ignoring...')
+            finally:
+                if self.worker_process.is_alive():
+                    logger.error("Worker is still alive, sending SIGKILL!")
+                    self.worker_process.kill()
+
+    def abort_grading(self) -> None:
+        assert self.worker_process_conn
+
+        try:
+            self.worker_process_conn.send((IPC.REQUEST_ABORT, ()))
+        except Exception:
+            logger.exception("Failed to send abort request to worker, did it race?")
+
+        self.stop()
+
+    def _worker_process_main(
+        self,
+        judge_process_conn: 'multiprocessing.connection.Connection',
+        worker_process_conn: 'multiprocessing.connection.Connection',
+    ) -> None:
+        """
+        Main body of judge worker process, which handles grading and sends grading results to the judge controller via
+        IPC.
+        """
+        worker_process_conn.close()
+        setproctitle(multiprocessing.current_process().name)
+
+        def _ipc_recv_thread_main() -> None:
+            """
+            Worker thread that listens for incoming IPC messages from the judge controller.
+            """
+            while True:
+                try:
+                    ipc_type, data = judge_process_conn.recv()
+                except:  # noqa: E722, whatever happened, we have to abort now.
+                    logger.exception("Judge unexpectedly hung up!")
+                    self._do_abort()
+                    return
+
+                if ipc_type == IPC.BYE:
+                    return
+                if ipc_type == IPC.REQUEST_ABORT:
```

Comment:
Will this look better as `elif`?
```suggestion
                elif ipc_type == IPC.REQUEST_ABORT:
```

Reply:
No opinion, so I made it an `elif`.